main/src/ca/uwaterloo/flix/language/errors/StratificationError.scala

```diff
@@ -17,14 +17,30 @@
 package ca.uwaterloo.flix.language.errors
 
 import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.TypedAst.Predicate.Body.{Filter, Loop, Negative, Positive}
+import ca.uwaterloo.flix.language.ast.TypedAst.Stratum
 import ca.uwaterloo.flix.language.ast._
+import ca.uwaterloo.flix.util.vt.VirtualString._
 import ca.uwaterloo.flix.util.vt.VirtualTerminal
 
 /**
   * An error raised to indicate that a property is violated.
   */
-case class StratificationError(constraints: List[TypedAst.Constraint]) extends CompilationError {
+case class StratificationError(stratum: List[Stratum], head: Symbol.TableSym, body: TypedAst.Predicate.Body, bodySym: Symbol.TableSym) extends CompilationError {
   val kind: String = "Stratification Error"
-  val source: SourceInput = constraints.head.loc.source
-  val message: VirtualTerminal = ??? // TODO
+  val source: SourceInput = head.loc.source
+  val message: VirtualTerminal = {
+    val vt = new VirtualTerminal
+    vt << Line(kind, source.format) << NewLine
+    vt << ">> Stratification Error in constraints" << NewLine
+    vt << "It was triggered by the rule: " << head.name << "(...) :- "
+    vt << (body match {
+      case Positive(sym, terms, loc) => sym.name + "(...)"
```

Comment:
Is there a way to pretty print source code? I'd like to properly print out the rule, but don't want to have to match on every type of pattern.